elide-graphql/src/main/java/com/yahoo/elide/graphql/GraphQLEntityProjectionMaker.java

```diff
@@ -0,0 +1,457 @@
+ * See LICENSE file in project root for terms.
+ */
+package com.yahoo.elide.graphql;
+
+import com.yahoo.elide.ElideSettings;
+import com.yahoo.elide.core.EntityDictionary;
+import com.yahoo.elide.core.pagination.Pagination;
+import com.yahoo.elide.core.sort.Sorting;
+import com.yahoo.elide.graphql.containers.ConnectionContainer;
+import com.yahoo.elide.graphql.containers.EdgesContainer;
+import com.yahoo.elide.request.Argument;
+import com.yahoo.elide.request.Attribute;
+import com.yahoo.elide.request.EntityProjection;
+
+import org.antlr.v4.runtime.ANTLRInputStream;
+import org.antlr.v4.runtime.CommonTokenStream;
+import org.atteo.evo.inflector.English;
+
+import graphql.GraphQL;
+import graphql.parser.antlr.GraphqlBaseVisitor;
+import graphql.parser.antlr.GraphqlLexer;
+import graphql.parser.antlr.GraphqlParser;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Supplier;
+
+/**
+ * A strategy that mapps a read-only GraphQL query to a collection of {@link EntityProjection}s.
+ * <p>
+ * <b>{@link GraphQLEntityProjectionMaker} produces {@link EntityProjection} that is only used in read operations</b>.
+ * Any write operations will result in runtime exceptions, although they might be supported in future releases. GraphQL
+ * query involving write operations should be processed by GraphQL API instead({@link GraphQL#execute(String)}).
+ * <p>
+ * {@link GraphQLEntityProjectionMaker} is not thread-safe and its concurrent access must be guarded by external
+ * synchronizations.
+ * <p>
+ * Caller should not invoke any {@link GraphQLEntityProjectionMaker} methods except for its constructor and
+ * {@link #make(String)}.
+ *
+ * @see EntityProjection
+ * @see ModelBuilder
+ * @see <a href="http://elide.io/pages/guide/11-graphql.html#sorting">Sorting</a>
+ *
+ * @// TODO: 2019-08-07 Support fragment query and update elide doc to include example of frangment query
+ */
+@Slf4j
+@RequiredArgsConstructor
+public class GraphQLEntityProjectionMaker extends GraphqlBaseVisitor<Void> {
+
+    /**
+     * All EntityProjections transformed from a GraphQL query.
+     * <p>
+     * GraphQL can query multiple entities in a single query, each root entity is transformed in an EntityProjections in
+     * this list.
+     */
+    @Getter(AccessLevel.PRIVATE)
+    private final List<EntityProjection> allEntityProjections = new ArrayList<>();
+
+    /**
+     * A map from EntityProjection.type to the EntityProjection.
+     */
+    @Getter(AccessLevel.PRIVATE)
+    private final Map<Class<?>, EntityProjection> projectionsByType = new HashMap<>();
```

Comment:
I don't think it is safe to a type will show up uniquely once in a query. We have to assume that a type can be asked for multiple times in a single request - each which a different projection.